arch/arm64/core/smp.c

```diff
@@ -72,40 +73,49 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 	master_core_mpid = MPIDR_TO_CORE(GET_MPIDR());
 
 	cpu_count = ARRAY_SIZE(cpu_node_list);
+
+#ifdef CONFIG_ARM64_FALLBACK_ON_RESERVED_CORES
+	__ASSERT(cpu_count >= CONFIG_MP_MAX_NUM_CPUS,
+		"The count of CPU Core nodes in dts is not greater or equal to CONFIG_MP_MAX_NUM_CPUS\n");
+#else
 	__ASSERT(cpu_count == CONFIG_MP_MAX_NUM_CPUS,
 		"The count of CPU Cores nodes in dts is not equal to CONFIG_MP_MAX_NUM_CPUS\n");
-
-	for (i = 0, j = 0; i < cpu_count; i++) {
-		if (cpu_node_list[i] == master_core_mpid) {
-			continue;
-		}
-		if (j == cpu_num - 1) {
-			cpu_mpid = cpu_node_list[i];
-			break;
-		}
-		j++;
-	}
-	if (i == cpu_count) {
-		printk("Can't find CPU Core %d from dts and failed to boot it\n", cpu_num);
-		return;
-	}
+#endif
 
 	arm64_cpu_boot_params.sp = Z_KERNEL_STACK_BUFFER(stack) + sz;
 	arm64_cpu_boot_params.fn = fn;
 	arm64_cpu_boot_params.arg = arg;
 	arm64_cpu_boot_params.cpu_num = cpu_num;
 
-	barrier_dsync_fence_full();
+	for (; i < cpu_count; i++) {
+		if (cpu_node_list[i] == master_core_mpid) {
+			continue;
+		}
 
-	/* store mpid last as this is our synchronization point */
-	arm64_cpu_boot_params.mpid = cpu_mpid;
+		cpu_mpid = cpu_node_list[i];
 
-	sys_cache_data_invd_range((void *)&arm64_cpu_boot_params,
-				  sizeof(arm64_cpu_boot_params));
+		barrier_dsync_fence_full();
 
-	if (pm_cpu_on(cpu_mpid, (uint64_t)&__start)) {
-		printk("Failed to boot secondary CPU core %d (MPID:%#llx)\n",
-		       cpu_num, cpu_mpid);
+		/* store mpid last as this is our synchronization point */
+		arm64_cpu_boot_params.mpid = cpu_mpid;
+
+		sys_cache_data_invd_range((void *)&arm64_cpu_boot_params,
+					  sizeof(arm64_cpu_boot_params));
+
+		if (pm_cpu_on(cpu_mpid, (uint64_t)&__start)) {
+			printk("Failed to boot secondary CPU core %d (MPID:%#llx)\n",
+			       cpu_num, cpu_mpid);
+#ifdef CONFIG_ARM64_FALLBACK_ON_RESERVED_CORES
+			continue;
+#else
+			return;
```

Comment:
I don't have a mean to test this but reading your words my impression was that when the fallback is not enabled and a core fails to boot the execution on the core just keeps looping in `start_cpu()`? If this is the case I would prefer to inform the user that the core failed to boot and panic instead of having the execution just stuck in the dark. Is my understanding correct?

This is honestly a bit of a weird situation where a failure in booting a secondary core is resulting in the booting core to be stuck.

Reply:
Your understanding is correct.  On today's mainline if a core fails to boot then the primary core gets stuck in start_cpu.  A message is printed to the user via printk in arch_start_cpu that the core failed to boot, but today execution gets stuck, a panic is not generated.

So sounds like you would like a panic added.  I can add that.

Reply:
Panics have been added.